modules/javafx.base/src/test/java/test/javafx/beans/value/ObservableValueFluentBindingsTest.java

```diff
@@ -0,0 +1,700 @@
+                        void shouldNoLongerBeCalled() {
+                            assertNothingIsObserved();
+
+                            property.set("Left");
+
+                            assertNothingIsObserved();
+                        }
+
+                        @Test
+                        void shouldNoLongerBeStronglyReferenced() {
+                            ReferenceAsserts.testIfNotStronglyReferenced(observableValue, () -> observableValue = null);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    @Nested
+    class When_orElse_CalledReturns_ObservableValue_Which {
+        private ObservableValue<String> observableValue = property.orElse("Empty");
+
+        @Test
+        void shouldNotBeNull() {
+            assertNotNull(observableValue);
+        }
+
+        @Test
+        void shouldNotBeStronglyReferenced() {
+            ReferenceAsserts.testIfNotStronglyReferenced(observableValue, () -> observableValue = null);
+        }
+
+        @Nested
+        class When_getValue_Called {
+            @Test
+            void shouldReturnPropertyValuesWithOperationApplied() {
+                assertEquals("Initial", observableValue.getValue());
+
+                property.set(null);
+
+                assertEquals("Empty", observableValue.getValue());
+            }
+        }
+
+        @Nested
+        class WhenObserved {
+            {
+                startObserving(observableValue);
+            }
+
+            @Test
+            void shouldApplyOperation() {
+                assertNothingIsObserved();
+
+                property.set("Left");
+
+                assertObserved("Left");
+
+                property.set(null);
+
+                assertObserved("Empty");
+            }
+
+            @Test
+            void shouldBeStronglyReferenced() {
+                ReferenceAsserts.testIfStronglyReferenced(observableValue, () -> observableValue = null);
+            }
+
+            @Nested
+            class AndWhenUnobserved {
+                {
+                    stopObserving(observableValue);
+                }
+
+                @Test
+                void shouldNoLongerBeCalled() {
+                    assertNothingIsObserved();
+
+                    property.set("Left");
+
+                    assertNothingIsObserved();
+                }
+
+                @Test
+                void shouldNoLongerBeStronglyReferenced() {
+                    ReferenceAsserts.testIfNotStronglyReferenced(observableValue, () -> observableValue = null);
+                }
+            }
+        }
+    }
+
+    @Nested
+    class When_flatMap_Called {
+        @Nested
+        class WithNull {
+            @Test
+            void shouldThrowNullPointerException() {
+                assertThrows(NullPointerException.class, () -> property.flatMap(null));
+            }
+        }
+        // TODO test for when something is flatMapped to null in getValue call
```

Comment:
Is this still relevant? I think that you tested it in line 407 `shouldIgnoreFlatMapsToNull`.

What I would like to see is a test when `left`, `right` or `unknown`'s value is set to `null`. What I see is `property`'s value being set to `null`, or the reference of one of the above being set to `null`, but not the value itself. Only when you compose `orElse` on `flatMap` this case is tested (line 604).
Is this what you meant?

Reply:
I left that TODO in by accident after my clean up of the test case, I added the case then.  I've however also added one more case that you described above:

                @Test
                void shouldObserveNullWhenFlatMappedPropertyIsSetToNull() {
                    property.set("Right");

                    assertObserved("RIGHT");

                    property.set(null);

                    assertObserved((String)null);
                }